From 47333b48bd2544b6b616ceba4b5ef15ed567c1da Mon Sep 17 00:00:00 2001
From: michaedw in build chroot <build@ctbu-bld5.cisco.com>
Date: Wed, 9 Mar 2011 07:42:19 +0000
Subject: [PATCH] getaddrinfo() result sorting deviates from RFC3484

2010-04-06  Ulrich Drepper  <drepper@redhat.com>

	* sysdeps/posix/getaddrinfo.c (default_scopes): Assign global
	scope to RFC 1918 addresses.
	* posix/gai.conf: Document difference from RFC 3484.
---
 posix/gai.conf              |   17 +++++++++++++----
 sysdeps/posix/getaddrinfo.c |    2 ++
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/posix/gai.conf b/posix/gai.conf
index 195287e..55e49bf 100644
--- a/posix/gai.conf
+++ b/posix/gai.conf
@@ -41,7 +41,7 @@
 #
 # precedence  <mask>   <value>
 #    Add another rule to the RFC 3484 precedence table.  See section 2.1
-#    and 10.3 in RFC 3484.  The default is:
+#    and 10.3 in RFC 3484.  The RFC requires:
 #
 #precedence  ::1/128       50
 #precedence  ::/0          40
@@ -56,9 +56,7 @@
 #
 # scopev4  <mask>  <value>
 #    Add another rule to the RFC 3484 scope table for IPv4 addresses.
-#    By default the scope IDs described in section 3.2 in RFC 3484 are
-#    used.  Changing these defaults should hardly ever be necessary.
-#    The defaults are equivalent to:
+#    The definitions in RFC 3484 are equivalent to:
 #
 #scopev4 ::ffff:169.254.0.0/112  2
 #scopev4 ::ffff:127.0.0.0/104    2
@@ -75,3 +73,14 @@
 #scopev4 ::ffff:169.254.0.0/112  2
 #scopev4 ::ffff:127.0.0.0/104    2
 #scopev4 ::ffff:0.0.0.0/96       14
+#
+#    For sites which use site-local IPv4 addresses behind NAT there is
+#    the problem that even if IPv4 addresses are preferred they do not
+#    have the same scope and are therefore not sorted first.  To change
+#    this use only these rules:
+#
+#scopev4 ::ffff:169.254.0.0/112  2
+#scopev4 ::ffff:127.0.0.0/104    2
+#scopev4 ::ffff:0.0.0.0/96       14
+#
+#    This is what the Red Hat / Ubuntu setting currently uses.
diff --git a/sysdeps/posix/getaddrinfo.c b/sysdeps/posix/getaddrinfo.c
index c61c72a..4a0b6b1 100644
--- a/sysdeps/posix/getaddrinfo.c
+++ b/sysdeps/posix/getaddrinfo.c
@@ -1101,10 +1101,12 @@ static const struct scopeentry
     /* Link-local addresses: scope 2.  */
     { { { 169, 254, 0, 0 } }, htonl_c (0xffff0000), 2 },
     { { { 127, 0, 0, 0 } }, htonl_c (0xff000000), 2 },
+#if 0
     /* Site-local addresses: scope 5.  */
     { { { 10, 0, 0, 0 } }, htonl_c (0xff000000), 5 },
     { { { 172, 16, 0, 0 } }, htonl_c (0xfff00000), 5 },
     { { { 192, 168, 0, 0 } }, htonl_c (0xffff0000), 5 },
+#endif
     /* Default: scope 14.  */
     { { { 0, 0, 0, 0 } }, htonl_c (0x00000000), 14 }
   };
-- 
1.7.0.4

